spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DistriParameterSynchronizer.scala

```diff
@@ -199,249 +170,347 @@ class BlockManagerParameterSynchronizer[T: ClassTag](partitionID: Int,
+      new ConcurrentHashMap[Int, CompressedTensor[T]](),
+      new ConcurrentHashMap[Int, Tensor[T]](), weights, grads))
   }
 
-  private def getParameterBlockId(name: String, counter: Int, pidFrom: Int, pidTo: Int): BlockId = {
-    SparkExtension.getLocalBlockId(name + counter +  pidFrom + "paraBytes" + pidTo)
+  override def put(name: String): Unit = {
+    val syncMeta = syncMetaMap.get(name)
+    syncMeta.counter += 1
+    val asyncTask = new AsyncTask(syncMeta.grads)
+    val futureTask = new FutureTask[Tensor[T]](asyncTask)
+    val syncRequest = new SyncRequest(new AtomicInteger(0), syncMeta, futureTask, asyncTask)
+    asyncTaskWaitingQueue.add(syncRequest)
+    if (syncMeta.counter > 1) {
+      val clearTask = new ClearTask(name, syncMeta.counter - 1,
+        partitionID, syncMeta.partitionToCount)
+      clearPool.execute(clearTask)
+    }
+    syncResults.put(name, futureTask)
+  }
+
+  override def get(name: String): (Tensor[T], Tensor[T]) = {
+    val syncMeta = syncMetaMap.get(name)
+    // no need to do aggregation for first forward
+    if (syncMeta.counter == 0) {
+      return (null, null)
+    }
+    require(syncResults.contains(name), "put must be done before get")
+    val res = syncResults.get(name).get.get()
+    (syncMeta.weights, res)
   }
 
   override def clear(): Unit = {
     shutdown = true
     longRunningThreads.asScala.foreach(_.interrupt())
     clearPool.shutdown
-    syncPool.shutdown
     workerPool.shutdown
+    syncPool.shutdown
     fetchPool.shutdown
     fetchCompletionPool.shutdown
   }
+
+  private class SyncRequest(var state: AtomicInteger, val syncMeta: SyncMeta[T],
+                            val futureTask: FutureTask[Tensor[T]],
+                            val asyncTask: AsyncTask)
+    extends Comparable[SyncRequest] {
+    override def compareTo(o: SyncRequest): Int = {
+      o.syncMeta.priority.compareTo(this.syncMeta.priority)
+    }
+  }
+
+  class AsyncTask(val parameter: Tensor[T]) extends Callable[Tensor[T]] {
+    override def call(): Tensor[T] = {
+      parameter
+    }
+  }
+
+
+  private class ClearTask(name: String, counter: Int, partitionID: Int,
+                          partitionToCount: Int)
+    extends Runnable {
+    override def run(): Unit = {
+      try {
+        (0 until partitionToCount).foreach(pid => {
+          val parameterBlockId = getParameterBlockId(name,
+            counter, partitionID, pid)
+          BlockManagerWrapper.removeBlock(parameterBlockId)
+        })
+        // only local parititon < partitionToCount, there are aggregated blocks
+        if (partitionID < partitionToCount) {
+          val aggregatedParameterBlockId = getParameterBlockId(s"${name}_aggregated",
+            counter, partitionID, -1)
+          BlockManagerWrapper.removeBlock(aggregatedParameterBlockId)
+        }
+      } catch {
+        case e: Exception =>
+          logger.info("exit thread gracefully")
```

Comment:
Why is it graceful when exiting in the exception ?

Reply:
because here it will catch the interrupted exception from the main thread